exporters/exporter.go

```diff
@@ -0,0 +1,72 @@
+package exporters
+
+import (
+	"github.com/algorand/go-algorand/agreement"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+)
+
+// ExporterConfig is a generic string which can be deserialized by each individual Exporter.
+type ExporterConfig string
+
+// ExportData is the interface which all data types sent to Exporters should implement
+type ExportData interface {
+	Round() uint64
+}
+
+// BlockExportData is provided to the Exporter on each round.
+type BlockExportData struct {
+	// Block is the block data written to the blockchain.
+	Block bookkeeping.Block
+
+	// Delta contains a list of account changes resulting from the block. Processor plugins may have modify this data.
+	Delta ledgercore.StateDelta
+
+	// Certificate contains voting data that certifies the block. The certificate is non deterministic, a node stops collecting votes once the voting threshold is reached.
+	Certificate agreement.Certificate
+}
+
+// Round returns the round to which the BlockExportData corresponds
+func (blkData *BlockExportData) Round() uint64 {
+	return uint64(blkData.Block.Round())
+}
+
+// ExporterMetadata contains fields relevant to identification and description of plugins.
+type ExporterMetadata struct {
+	Name        string
+	Description string
+	Deprecated  bool
+}
+
+// Exporter defines the interface for plugins
+type Exporter interface {
+	// Metadata associated with each Exporter.
+	Metadata() ExporterMetadata
+
+	// Connect will be called during initialization, before block data starts going through the pipeline.
+	// Typically used for things like initializing network connections.
+	// The ExporterConfig passed to Connect will contain the Unmarhsalled config file specific to this plugin.
+	// Should return an error if it fails--this will result in the Indexer process terminating.
+	Connect(cfg ExporterConfig) error
+
+	// Config returns the configuration options used to create an Exporter.
+	// Initialized during Connect, it should return nil until the Exporter has been Connected.
+	Config() ExporterConfig
+
+	// Disconnect will be called during termination of the Indexer process.
+	// There is no guarantee that plugin lifecycle hooks will be invoked in any specific order in relation to one another.
+	// Returns an error if it fails which will be surfaced in the logs, but the process is already terminating.
+	Disconnect() error
+
+	// Receive is called for each block to be processed by the exporter.
+	// Should return an error on failure--retries are configurable.
+	Receive(exportData ExportData) error
```

Comment:
The main part I'm not sure about is whether we should have this abstraction in the first iteration. It means there will be a required type check in all plugins to convert it into `BlockExportData`.

I think it's ok to leave it like this for now, but lets keep an eye on it.